### PR TITLE
Fix BIO_dgram_pair buffer arithmetic bug

### PR DESCRIPTION
**Pull request by hlandau**
Managed to reproduce the test segfault. This appears to be the cause of the CI failures.

### CODE REVIEW

**Comment by hlandau** [comment]
It is a bit unfortunate we don't dump the OPENSSL_TEST_RAND_ORDER value if a test binary segfaults.

**Review by t-j-h — APPROVED** [review]

**Comment by paulidale** [comment]
We dump the seed in the verdict function.  Perhaps we should dump it when setting it instead?

**Review by paulidale — APPROVED** [review]
Sigh, this is the kind of thing I'd expect static analysis to be able to pick up.

**Comment by hlandau** [comment]
@paulidale Yeah, it was surprising how ubsan only picked this up stochastically.

**Comment by tom-cosgrove-arm** [comment]
Hasn't this already been done in #19284, which has been merged?

File: crypto/bio/bss_dgram_pair.c [hunk 1]
```diff
@@ -745,7 +745,9 @@ static size_t dgram_pair_read_inner(struct bio_dgram_pair_st *b, uint8_t *buf, s
 
         ring_buf_pop(&b->rbuf, src_len);
 
-        buf         += src_len;
+        if (buf != NULL)
```
Comment (tom-cosgrove-arm) [inline crypto/bio/bss_dgram_pair.c:748]: (v minor) might be a bit nicer to move this into the existing `if (buf != NULL)` check at line 743

**Comment by paulidale** [comment]
I was more thinking about Coverity and friends.  Earlier null check, so buf could be null, then adding a value to null is crazy.

**Comment by hlandau** [comment]
Closing as made redundant by https://github.com/openssl/openssl/pull/19284.

**Pull request closed without merging**